### ISSUE TITLE
Bug: Inlay hint type annotation for includes_bytes! macro displays size 0 rather than the correct size

### ISSUE THREAD

**Issue by budde25** [rust-lang/rust-analyzer#12181]
The type hint for a variable that is the result of the include_bytes! macro shows size zero rather than its correct size (which is known at compile time)

<img width="791" alt="Screen Shot 2022-05-06 at 2 19 02 PM" src="https://user-images.githubusercontent.com/16231105/167203714-1e9b7803-0379-4c1d-af56-8b6732735b2a.png">

However after typing the exact hint in as it is presented, RA shows an error indicating a type mismatch on the size of the array.

<img width="949" alt="image" src="https://user-images.githubusercontent.com/16231105/167203838-971c3cd9-d599-4f7d-9819-8b22554dc389.png">

I would expect the hint to display `: &[u8; 122880]`.

**rust-analyzer version**: (eg. output of "Rust Analyzer: Show RA Version" command)
rust-analyzer version: 5dce1ff02 2022-05-02 stable

**rustc version**: (eg. output of `rustc -V`)
rustc 1.60.0 (7737e0b5c 2022-04-04)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

**Comment by lukechu10** [comment]
The error message comes from `rustc`, not `rust-analyzer`. I believe this is an issue with how the macro is expanded right here:
https://github.com/rust-lang/rust-analyzer/blob/81b3e6d124db18cd6015b7d763addfd63e0bd54a/crates/hir-expand/src/builtin_fn_macro.rs#L579-L587

**Comment by bjorn3** [comment]
> The type hint for a variable that is the result of the include_bytes! macro shows size zero rather than its correct size (which is known at compile time)

Rust-analyzer expands `include_bytes!()` as `&[]` which indeed has length zero. Most of the time it doesn't matter, and it improves performance for large files.

**Comment by flodiebold** [comment]
I think this will be hard to fix, since we don't even have non-rs files in the VFS, so even if we wanted to, loading them would not be easy.

**Comment by bjorn3** [comment]
Would it be possible to expand `include_bytes!()` into something of type `&[u8; _]` without giving any specific size? Or if the file is loaded into the vfs give a specific size, but no actual value. For example `loop {} as &[u8; _]` or `loop {} as &[u8; 42]` in case of a known size.

**Comment by flodiebold** [comment]
Yeah, I was considering that as well. I didn't realize that `&[u8; _]` is actually valid (if unstable) Rust though, so yeah, that would at least be an improvement over the current situation. Not sure whether `loop {}` is a good idea though, because of divergence tracking :thinking:

**Comment by bjorn3** [comment]
`(|| loop {})()` does the same except without being considered unreachable, right?

**Comment by HKalbasi** [comment]
Just a `[0; _]` without `loop {}` does the job I think.

**Comment by flodiebold** [comment]
I did not think `_` would be allowed in the repeat expression, but apparently it is!

**Comment by HKalbasi** [comment]
> The remaining question would be how we currently handle that in rust-analyzer.

It would behave as an unevaluatable constant, we will try to infer it from the rest of the program.

**Comment by budde25** [comment]
> I think this will be hard to fix, since we don't even have non-rs files in the VFS, so even if we wanted to, loading them would not be easy. (Plus we might need to be careful about performance when we suddenly have tokens of 100s of KB or even bigger in the AST.)

I'm not sure how it works on other platforms but isn't it possible see file size without loading the whole file into memory? https://man7.org/linux/man-pages/man7/inode.7.html although that might be beyond the scope of RA, especially if it is platform specific.

for example using stat gives the correct size
```
stat crates/tegra-rcm/src/test/hekate_ctcaer_5.7.2_ref_payload.bin 
  File: crates/tegra-rcm/src/test/hekate_ctcaer_5.7.2_ref_payload.bin
  Size: 122880          Blocks: 240        IO Block: 4096   regular file
Device: 0,40    Inode: 1686053     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/    budd)   Gid: ( 1000/    budd)
Context: unconfined_u:object_r:unlabeled_t:s0
Access: 2022-05-07 15:37:01.570224647 -0500
Modify: 2022-05-06 14:59:04.683658031 -0500
Change: 2022-05-06 14:59:04.683658031 -0500
 Birth: 2022-05-06 14:59:04.683658031 -0500
```
